blaze-client/src/main/scala/org/http4s/client/blaze/PoolManager.scala

```diff
@@ -1,67 +1,140 @@
-      if(!closed) {
-        closed = true
-        cs.foreach(_.stage.shutdown())
+  private def stats =
+    s"allocated=${allocated} idleQueue.size=${idleQueue.size} waitQueue.size=${waitQueue.size}"
+
+  private def createConnection(key: RequestKey): Unit = {
+    if (allocated < maxTotal) {
+      allocated += 1
+      logger.debug(s"Creating connection: ${stats}")
+      Task.fork(builder(key)).runAsync {
+        case \/-(stage) =>
+          logger.debug(s"Submitting fresh connection to pool: ${stats}")
+          actor ! ReturnConnection(key, stage)
+        case -\/(t) =>
+          logger.error(t)("Error establishing client connection")
+          actor ! DisposeConnection(key, None)
       }
     }
+    else {
+      logger.debug(s"Too many connections open.  Can't create a connection: ${stats}")
+    }
   }
 
-  override def recycleClient(request: Request, stage: BlazeClientStage): Unit =
-    cs.synchronized {
-      if (closed) stage.shutdown()
-      else {
-          logger.debug("Recycling connection.")
-          cs += Connection(request.uri.scheme, request.uri.authority, stage)
-
-        while (cs.size >= maxPooledConnections) {  // drop connections until the pool will fit this connection
-          logger.trace(s"Shutting down connection due to pool excess: Max: $maxPooledConnections")
-          val Connection(_, _, stage) = cs.dequeue()
-          stage.shutdown()
+  private val actor = Actor[Protocol] {
+    case RequestConnection(key, callback) =>
+      logger.debug(s"Requesting connection: ${stats}")
+      if (!isClosed) {
+        def go(): Unit = {
+          idleQueue.dequeueFirst(_.requestKey == key) match {
+            case Some(stage) if !stage.isClosed =>
+              logger.debug(s"Recycling connection: ${stats}")
+              callback(stage.right)
+            case Some(closedStage) =>
+              logger.debug(s"Evicting closed connection: ${stats}")
+              allocated -= 1
+              go()
+            case None =>
+              logger.debug(s"No connections available.  Waiting on new connection: ${stats}")
+              createConnection(key)
+              waitQueue.enqueue(callback)
+          }
         }
+        go()
       }
-    }
+      else
+        callback(new IllegalStateException("Connection pool is closed").left)
 
-  override def getClient(request: Request, freshClient: Boolean): Task[BlazeClientStage] = Task.suspend {
-    cs.synchronized {
-      if (closed) Task.fail(new Exception("Client is closed"))
-      else if (freshClient)  {
-        logger.debug("Creating new connection, per request.")
-        builder(request)
+    case ReturnConnection(key, stage) =>
+      logger.debug(s"Reallocating connection: ${stats}")
+      if (!isClosed) {
+        if (!stage.isClosed) {
+          stage.reset()
+          if (waitQueue.nonEmpty) {
+            logger.debug(s"Fulfilling waiting connection request: ${stats}")
+            waitQueue.dequeue.apply(stage.right)
```

Comment:
Ugh.  Yes.

In the long term, individual keyed pools would stop us from having to shuffle through all the connections, and also permit a (common in other clients) cap on the number of open requests per key.